src/test/java/io/kestra/plugin/nats/ConsumeTest.java

```diff
@@ -0,0 +1,170 @@
+import io.nats.client.api.DeliverPolicy;
+import io.nats.client.impl.Headers;
+import jakarta.inject.Inject;
+import org.hamcrest.Matchers;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.time.ZonedDateTime;
+import java.time.temporal.ChronoUnit;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.*;
+
+class ConsumeTest extends NatsTest {
+    @Inject
+    protected RunContextFactory runContextFactory;
+
+    @Test
+    void consumeMessageFromSubject() throws Exception {
+        Connection connection = Nats.connect(Options.builder().server("localhost:4222").userInfo("kestra", "k3stra").build());
+
+        AtomicReference<Instant> messageInstant = new AtomicReference<>();
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+        JetStreamSubscription subscription = connection.jetStream().subscribe("kestra.consumeMessageFromSubject.topic", PullSubscribeOptions.builder()
+            .configuration(ConsumerConfiguration.builder().deliverPolicy(DeliverPolicy.New).build())
+            .build()
+        );
+
+        Executors.newSingleThreadExecutor().submit(Rethrow.throwRunnable(() ->
+            {
+                List<Message> messages = subscription.fetch(1, Duration.ofSeconds(2));
+                messageInstant.set(
+                    // Compulsory to match ION-serialized instant precision
+                    Instant.ofEpochMilli(messages.get(0).metaData().timestamp().toInstant().toEpochMilli())
+                );
+                countDownLatch.countDown();
+            }
+        ));
+
+        Headers headers = new Headers();
+        String expectedHeaderKey = "someHeaderKey";
+        String expectedHeaderValue = "someHeaderValue";
+        headers.add(expectedHeaderKey, expectedHeaderValue);
+        connection.publish("kestra.consumeMessageFromSubject.topic", headers, "Hello Kestra".getBytes());
+        connection.publish("kestra.consumeMessageFromSubject.anotherTopic", "Hello Again".getBytes());
+
+        Consume.Output output = Consume.builder()
+            .url("localhost:4222")
+            .username("kestra")
+            .password("k3stra")
+            .subject("kestra.consumeMessageFromSubject.>")
+            .durableId("consumeMessageFromSubject-" + UUID.randomUUID())
+            .deliverPolicy(DeliverPolicy.LastPerSubject)
+            .pollDuration(Duration.ofSeconds(1))
+            .build()
+            .run(runContextFactory.of());
+
+        List<Map<String, Object>> result = toMessages(output);
+
+        countDownLatch.await();
+        connection.close();
```

Comment:
Move all `connection` usage together for better readability. You can close the connection just after message publication (and create it just before)

Reply:
I'm also using the same connection to subscribe before publishing and the countDownLatch is my only way to know I can now close the connection. The only thing I can do is move up the close call inside the Executor ? Or is there something I'm missing ?

Reply:
OK